src/FunctionalExtras/Data/Validation.cs

```diff
@@ -0,0 +1,315 @@
+    /// <param name="first">The first <see cref="Validation"/>.</param>
+    /// <typeparam name="F">The underlying failure type.</typeparam>
+    /// <typeparam name="S">The underlying success type.</typeparam>
+    /// <returns>The first <code>Success</code> for two successes, the first <code>Failure</code> for mixed; otherwise,
+    /// a <code>Failure</code> of the concatenation of the failure values.</returns>
+    /// <exception cref="ArgumentNullException">if either <see cref="Validation"/> is <code>null</code>.</exception>
+    public static IValidation<F, S> Concat<F, S>(IValidation<F, S> second, IValidation<F, S> first)
+    {
+      RequireNonNull(second, "second validation must not be null");
+      RequireNonNull(first, "first validation must not be null");
+
+      IValidation<F, S> result = first;
+
+      if (first.IsSuccess() && second.IsFailure())
+      {
+        result = second;
+      }
+      else if (second.IsFailure())
+      {
+        result = Failure<F, S>(new List<IValidation<F, S>> { first, second }
+          .SelectMany(FromFailure<F, S>(Enumerable.Empty<F>()))
+          .ToList()
+        );
+      }
+
+      return result;
+    }
+
+    /// <summary>
+    /// Creates a <code>Failure</code> from an arbitrary value.
+    /// </summary>
+    /// <typeparam name="F">The underlying failure type.</typeparam>
+    /// <typeparam name="S">The underlying success type.</typeparam>
+    /// <param name="value">The value.</param>
+    /// <returns>A <code>Failure</code> of the value.</returns>
+    public static IValidation<F, S> Failure<F, S>(F value) => new Failure<F, S>(value);
+
+    /// <summary>
+    /// Creates a <code>Failure</code> from an arbitrary collection of values.
+    /// </summary>
+    /// <typeparam name="F">The underlying failure type.</typeparam>
+    /// <typeparam name="S">The underlying success type.</typeparam>
+    /// <param name="values">The values.</param>
+    /// <returns>A <code>Failure</code> of the values.</returns>
+    public static IValidation<F, S> Failure<F, S>(IEnumerable<F> values) => new Failure<F, S>(values);
+
+    /// <summary>
+    /// Extracts from a collection of <see cref="Validation"/> all of the <code>Failure</code> elements in extracted
+    /// order. The underlying collections are concatenated together.
+    /// </summary>
+    /// <typeparam name="F">The underlying failure type.</typeparam>
+    /// <typeparam name="S">The underlying success type.</typeparam>
+    /// <param name="enumerable">The enumerable of <see cref="Validation"/></param>
+    /// <returns>The enumerable of underlying <code>Failure</code> values.</returns>
+    public static IEnumerable<F> Failures<F, S>(IEnumerable<IValidation<F, S>> enumerable) =>
+      (enumerable ?? Enumerable.Empty<IValidation<F, S>>())
+        .Where(IsNotNull)
+        .Where(IsFailure)
+        .SelectMany(FromFailure<F, S>(Enumerable.Empty<F>()))
+        .ToList();
```

Comment:
Intentional `ToList` with `IEnumerable` return?

Reply:
Can I leave it at the `SelectMany` or do I need to terminate the Linq query somehow? I sorta took this idea from the Javas where you have to close out the stream with a `.collect` call.

Reply:
You can leave it with `SelectMany` unless you specifically need to execute the enumeration.

Reply:
Updated this and `Either` to do the same.